File: practice/algorithms/bit_manipulation/xor_key/xor_key.py
#!/usr/bin/env python

import bisect

class Node(object):
    def __init__(self):
        self._Left  = None
        self._Right = None
        self.Index = []
    
    def get_right(self):
        return self._Right
    
    def set_right(self, i_node):
        self._Right = i_node
    
    def get_left(self):
        return self._Left
    
    def set_left(self, i_node):
        self._Left = i_node
    
    def add_index(self, i_index):
        bisect.insort_right(self.Index, i_index)
    
    def has_index_in_range(self, i_min_index, i_max_index):
        min_pos = bisect.bisect_left(self.Index, i_min_index)
        max_pos = bisect.bisect_right(self.Index, i_max_index)
        return min_pos < max_pos
    
class Trie:
    def __init__(self, i_arr):
        self.Root = Node()
        self.NumBits = 15
        self.Mask = [1 << POS for POS in range(self.NumBits)]
        
        for POS in range(len(i_arr)):
            val = i_arr[POS]
            self.Root.add_index(POS)
            
            node = self.Root
            nb = self.NumBits - 1
            while nb >= 0:
                mask = self.Mask[nb]
                if val & self.Mask[nb] == 0:
                    if not node.get_left():
                        node.set_left(Node())
                    node = node.get_left()
                else:
                    if not node.get_right():
                        node.set_right(Node())
                    node = node.get_right()
                node.add_index(POS)
                nb -= 1
      
    def max_xor(self, i_value, i_min_index, i_max_index):
        if not self.Root.has_index_in_range(i_min_index, i_max_index):
            return 2 ** self.NumBits - 1
        
        res = 0
        node = self.Root
        nb = self.NumBits - 1
        while nb >= 0:
            mask = self.Mask[nb]
            if i_value & mask == 0:
                if node.get_right() and (not node.get_left() or node.get_right().has_index_in_range(i_min_index, i_max_index)):
                    res += mask
                    node = node.get_right()
                else:
                    node = node.get_left()
            elif node.get_left() and (not node.get_right() or node.get_left().has_index_in_range(i_min_index, i_max_index)):
                res += mask
                node = node.get_left()
            else:
                node = node.get_right()
            nb -= 1
        return res

t = int(raw_input().strip())
for _ in range(t):
    n, q = map(int, raw_input().strip().split())
    arr = map(int, raw_input().strip().split())
    trie = Trie(arr)
    for __ in xrange(q):
        a, p, q = map(int, raw_input().strip().split())
        print trie.max_xor(a, p - 1, q - 1)

File: practice/algorithms/bit_manipulation/xor_key/xor_key.cpp
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

struct Node
{
    Node()
        : left(nullptr)
        , right(nullptr)
    {}
    
    void add_index(int i_idx) { index.insert(
        upper_bound(index.begin(), index.end(), i_idx), i_idx); };
    
    bool has_index_in_range(int i_min_index, int i_max_index)
    {
        int min_pos = distance(index.begin(), 
                               lower_bound(index.begin(), index.end(), i_min_index));
        int max_pos = distance(index.begin(), 
                               upper_bound(index.begin(), index.end(), i_max_index));
        return min_pos < max_pos;
    }

    Node * left;
    Node * right;
    vector<int> index;
};

class Trie
{
public:
    Trie(const vector<int> & i_arr)
        : m_root(new Node())
        , m_num_bits(15)
        , m_mask(vector<int>(15))
    {
        for (int pos = 0; pos < m_num_bits; ++pos)
        {
            m_mask[pos] = 1 << pos;
        }
            
        for (int pos = 0; pos < i_arr.size(); ++pos)
        {
            int val = i_arr[pos];
            m_root->add_index(pos);
            
            Node * node = m_root;
            int nb = m_num_bits - 1;
            while (nb >= 0)
            {
                int mask = m_mask[nb];
                if ((val & mask) == 0)
                {
                    if (node->left == nullptr)
                    {
                        node->left = new Node();
                    }
                    node = node->left;
                }
                else
                {
                    if (node->right == nullptr)
                    {
                        node->right = new Node();
                    }
                    node = node->right;
                }
                node->add_index(pos);
                nb--;
            }
        }
    }
    
    int max_xor(int i_value, int i_min_index, int i_max_index) const
    {
        if (!m_root->has_index_in_range(i_min_index, i_max_index))
        {
            return 1 << m_num_bits - 1;
        }
        
        int res = 0;
        Node * node = m_root;
        int nb = m_num_bits - 1;
        while (nb >= 0)
        {
            int mask = m_mask[nb];
            if ((i_value & mask) == 0)
            {
                if (node->right != nullptr && 
                    (node->left == nullptr || (node->right)->has_index_in_range(i_min_index, i_max_index)))
                {
                    res += mask;
                    node = node->right;
                }
                else
                {
                    node = node->left;
                }
            }
            else if (node->left != nullptr && 
                     (node->right == nullptr || (node->left)->has_index_in_range(i_min_index, i_max_index)))
            {
                res += mask;
                node = node->left;
            }
            else
            {
                node = node->right;
            }
            nb--;
        }
        
        return res;
    }
    
private:
    Node * m_root;
    int m_num_bits;
    vector<int> m_mask;
};

int main()
{
    int t, n, q;
    int a, l ,r;
    
    cin >> t;
    while(t--)
    {
        cin >> n >> q;
        
        vector<int> arr(n);
        for (int pos = 0; pos < n; ++pos)
        {
            cin >> arr[pos];
        }
        
        Trie trie(arr);
        while (q--)
        {
            cin >> a >> l >> r;
            cout << trie.max_xor(a, l - 1, r - 1) << endl;
        }
    }
    return 0;
}

